lib/tests/test_core.c

```diff
@@ -48,12 +48,82 @@ test_strerror_tskit(void)
     }
 }
 
+static void
+test_probability_list_select(void)
+{
+    double short_one = 1.0 - 0.5 * DBL_EPSILON;
```

Comment:
Kind of related, I think, but I think I should make the double const so that it's more clear this an alias for a value.

What is the thinking behind not mixing declarations and code? Just avoiding expressing too many ideas on a single line or something else?

Reply:
Aside from being an organising principal, older C standards (and thus compilers) don't allow mixed code/declarations. A compiler has to generate low-level code that moves the stack pointer for each variable declared on the stack. The old way assumed all variables were declared together in one block, so the compiler could generate a single instruction to move the stack pointer once, instead of separate instructions for each variable. In addition, at the start of a function, the compiler is already shifting the stack pointer because the stack is a commonly used for passing function arguments. The newer standards (and compilers) just move the stack pointer again when there are mixed code/declarations (or maybe some compilers reorder instructions to avoid this?).

Reply:
The thinking originally was to keep older compilers happy. Since we've fixed on C99, this probably isn't an issue now. 

One of the key reasons for keeping all the declarations at the top of a function is making sure that this idiom is always guaranteed to work:
```
double *some_pointer = NULL;

// later
some_pointer = malloc(...);
if (some_pointer == NULL) {
    ret = no memory error
    goto out;
}
// still later
out:
     tsk_safe_free(some_pointer);
     return ret;
```

making sure that all the pointers are initialised at the top of the function makes sure that we're always guaranteed that there's a valid pointer when we take the error path. This is something we really use heavily in the Python-C layer in particular.

For other variables it's a good sign your function is getting too complicated if you've got loads of variables declared at the top.

Reply:
In older C I believe this kind of thing would not be uncommon, many standard headers files are full of similar:
```
#define SHORT_ONE (1.0 - 0.5 * DBL_EPSILON)
```
in contrast to
```
double const short_one = 1.0 - 0.5 * DBL_EPSILON;
```
Is the first line not mixing declaration vs code? whereas the 2nd one is?

Does the first one make @jeromekelleher grumble but not the other? If so I'm curious why.

Reply:
I think it's fine here @castedo - this is testing code which gets pretty ugly anyway. I would suggest different approaches for production code.

Reply:
We're targetting a much narrower set of compilers here too, since it's only developers that'll be running the C tests. We don't have to worry about things working on Windows for example.